goethe/corpora.py

```diff
@@ -1,168 +1,52 @@
-        strides = a.strides + (a.strides[-1],)
-        return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
-
-    def number_ngrams(self, n):
-        for numbers in self.number_sentences():
-            for ngram in self.rolling_window(np.array(numbers), n):
-                yield ngram
-
-    def number_ngrams_(self, n):
-        for numbers in self.number_sentences():
-            for i in range(0, len(numbers) - n + 1):
-                yield numbers[i:i+n]
-
-    def sentences(self, words=True):
-        """Iterate over lines in files returning sentences.
+        path = self.file_path(use_tokens=True)
+        with open(path) as f:
+            yield from (l.strip().split() for l in f)
+
+    def file_path(self, use_tokens):
+        """Return path to either sentence or token file.
+        Example:
+            'abc/def/corpus'
+                --> 'abc/def/corpus/corpus.txt'
+                --> 'abc/def/corpus/corpus.tokens.txt'
         """
-        files = self.sentence_files()
-        trans = str.maketrans('', '', self.tokenizer_filter)
-        for f in files:
-            for line in open(f):
-                # Lines are of form: 'LineNumber\tActualSentence\n'
-                sent = line.split('\t')[1].strip().lower().translate(trans)
-                yield word_tokenize(sent) if words else sent
-
-    def sentences_n(self, n):
-        for i, s in enumerate(self.sentences()):
-            if self.max_sentences and i >= n:
-                break
-            yield s
+        corpus_name = os.path.basename(self.path)
+        file = ('%s.tokens.txt' if use_tokens else '%s.txt') % corpus_name
+        return os.path.join(self.path, file)
 
-    def sentence_files(self):
-        """Find all sentence files in 'dirname'.
-        Assumes file tree like following:
-            dirname/
-                deu_news_2015_3M/
-                    deu_news_2015_3M-sentences.txt
-                    ...
-                deu_wikipedia_2014_3M/
-                    deu_wikipedia_2014_3M-sentences.txt
-                    ...
-                ...
+    def random(self, k, use_tokens=None):
+        """Randomly select a list of k elements from either tokens or sents.
+        (Will load k elements into memory!)
         """
-        files = []
-        for corpus in os.listdir(self.dir):
-            # Select only corpora with 'self.lang' language
-            if not corpus.startswith(self.lang):
-                continue
-
-            if self.corpus_name and corpus != self.corpus_name:
-                continue
-
-            for f in os.listdir(os.path.join(self.dir, corpus)):
-                if not f.endswith('sentences.txt'):
-                    continue
-                if self.name_filter and not self.name_filter(f):
-                    continue
-                sentences_file = os.path.join(self.dir, corpus, f)
-                files.append(sentences_file)
-        return files
+        iterator = (self
```

Comment:
Or we add the possibility to limit both methods without any sampling, because it may not be suitable to hold all the data in the main memory. Then the user must explicitly use the random method to get shuffled data.